src/util/messagepipe.h

```diff
@@ -1,109 +1,83 @@
-                        sender_fifo_size, receiver_fifo_size));
-
-        return {
-                std::make_unique<MessagePipe<SenderMessageType, ReceiverMessageType>>(
-                        pipe->m_receiver_messages,
-                        pipe->m_sender_messages,
-                        new ReferenceHolder<TwoWayMessagePipe<SenderMessageType,
-                                ReceiverMessageType>>(pipe)),
-                std::make_unique<MessagePipe<ReceiverMessageType, SenderMessageType>>(
-                        pipe->m_sender_messages,
-                        pipe->m_receiver_messages,
-                        new ReferenceHolder<TwoWayMessagePipe<SenderMessageType,
-                                ReceiverMessageType>>(pipe))};
-    }
-
-  private:
-    TwoWayMessagePipe(int sender_fifo_size, int receiver_fifo_size)
-            : m_receiver_messages(receiver_fifo_size),
-              m_sender_messages(sender_fifo_size) {
-    }
-
-    // Messages waiting to be delivered to the receiver.
-    rigtorp::SPSCQueue<SenderMessageType> m_receiver_messages;
-    // Messages waiting to be delivered to the sender.
-    rigtorp::SPSCQueue<ReceiverMessageType> m_sender_messages;
+/// makeTwoWayMessagePipe is a bare-bones wrapper around the above rigtorp::SPSCQueue class that
+/// facilitates non-blocking two-way communication. To keep terminology clear,
+/// there are two sides to the message pipe, the sender side and the receiver
+/// side. The non-blocking aspect of the underlying rigtorp::SPSCQueue class requires that the
+/// sender methods and target methods each only be called from a single thread
+/// The most common use-case of this class is sending and receiving messages
+/// with the callback thread without the callback thread blocking.
+/// Returns a pair of MessagePipes,
+/// the first is the sender's pipe (sends SenderMessageType and receives
+/// ReceiverMessageType messages) and the second is the receiver's pipe
+/// (sends ReceiverMessageType and receives SenderMessageType messages).
+template<class SenderMessageType, class ReceiverMessageType>
+static std::pair<
+        std::unique_ptr<MessagePipe<SenderMessageType, ReceiverMessageType>>,
+        std::unique_ptr<MessagePipe<ReceiverMessageType, SenderMessageType>>>
+makeTwoWayMessagePipe(
+        int sender_fifo_size,
+        int receiver_fifo_size) {
+    auto sender_pipe = std::make_shared<rigtorp::SPSCQueue<ReceiverMessageType>>(sender_fifo_size);
+    auto receiver_pipe =
+            std::make_shared<rigtorp::SPSCQueue<SenderMessageType>>(
+                    receiver_fifo_size);
 
-    DISALLOW_COPY_AND_ASSIGN(TwoWayMessagePipe);
+    return {
+            std::make_unique<MessagePipe<SenderMessageType, ReceiverMessageType>>(
+                    receiver_pipe,
+                    sender_pipe),
+            std::make_unique<MessagePipe<ReceiverMessageType, SenderMessageType>>(
+                    sender_pipe,
+                    receiver_pipe)};
```

Comment:
Not entirely sure, but I think you may be able to use `std::move` here

Reply:
How exactly? Can you share you suggestion?

Reply:
I've seen place in the code where you would see something like

```suggestion
            std::make_unique<MessagePipe<SenderMessageType, ReceiverMessageType>>(
                    std:move(receiver_pipe),
                    std:move(sender_pipe)),
            std::make_unique<MessagePipe<ReceiverMessageType, SenderMessageType>>(
                    std:move(sender_pipe),
                    std:move(receiver_pipe))};
```
but perhaps this is redundant?

Reply:
now that would actually be a bug afaict. `std::move`ing here will result in the `shared_ptr( shared_ptr&& r )` constructor being called. From cppreference:
> 10) Move-constructs a shared_ptr from r. After the construction, *this contains a copy of the previous state of r, r is empty and its stored pointer is null. 

So using either `sender_pipe`/`receiver_pipe` after moving from it would just leave it null. We can only ever move out of a variable once. 

We could do something like

```suggestion
            std::make_unique<MessagePipe<SenderMessageType, ReceiverMessageType>>(
                    receiver_pipe,
                    sender_pipe),
            std::make_unique<MessagePipe<ReceiverMessageType, SenderMessageType>>(
                    std::move(sender_pipe),
                    std::move(receiver_pipe))};
```

Which would result in only one move, but its undefined which argument here is evaluated first (so the value could be moved from before the actually desired copy took place). 

Something like this should work though:
```suggestion
        auto rec_to_sender = std::make_unique<MessagePipe<SenderMessageType, ReceiverMessageType>>(
                    receiver_pipe,
                    sender_pipe);
        auto sender_to_rec = std::make_unique<MessagePipe<ReceiverMessageType, SenderMessageType>>(
                    sender_pipe,
                    receiver_pipe)};
        return {std::move(rec_to_sender), std::move(sender_to_rec)};
```